### PR TITLE
Detect t.try() usage in hooks; experimentally disable t.snapshot()

### PR DESCRIPTION
**Pull request by okyantoro**
This change will disable snapshots and try assertions in hooks. But I am not sure with the unit test. Let me know if additional work is required.

Fixes https://github.com/avajs/ava/issues/2523

### CODE REVIEW

File: lib/assert.js [hunk 1]
```diff
@@ -690,6 +691,14 @@ class Assertions {
 		});
 
 		this.snapshot = withSkip((expected, ...rest) => {
+			if (disableSnapshots) {
+				fail(new AssertionError({
+					assertion: 'snapshot',
+					message: '`t.snapshot()` is not allowed in hooks',
```
Comment (novemberborn) [inline lib/assert.js:697]:
```suggestion
					message: '`t.snapshot()` cannot be used here',
```

File: lib/test.js [hunk 2]
```diff
@@ -40,7 +40,8 @@ class ExecutionContext extends assert.Assertions {
 				return test.compareWithSnapshot(options);
 			},
 			powerAssert: test.powerAssert,
-			experiments: test.experiments
+			experiments: test.experiments,
+			disableSnapshots: test.isHook === true
```
Comment (novemberborn) [inline lib/test.js:44]: This is a breaking change. We need to [add an experiment](https://github.com/avajs/ava/blob/master/lib/load-config.js#L10) for this and then _also_ check if that's been set in `test.experiments`.

File: test-tap/hooks.js [hunk 3]
```diff
@@ -566,3 +567,37 @@ test('teardowns cannot be used in hooks', async t => {
...
+		await t.try(tt => tt.pass());
+
+		t.pass();
+	});
+
+	const result = await instance.run();
+
+	t.notOk(result.passed);
+	t.match(result.error.savedError.message, /not allowed in hooks/);
+});
```
Comment (novemberborn) [inline test-tap/hooks.js:603]: I'd like to self-host these tests. See https://github.com/avajs/ava/tree/master/test/

I'm thinking a directory for "hooks" and then two tests in there? Have a look at the `test-timeouts` in https://github.com/avajs/ava/pull/2524/files.

File: lib/assert.js (hunk 1 above)
Comment (okyantoro) [reply lib/assert.js]: addressed

File: lib/test.js (hunk 2 above)
Comment (okyantoro) [reply lib/test.js:44]: Addressed

File: lib/assert.js [hunk 4]
```diff
@@ -690,6 +691,15 @@ class Assertions {
 		});
 
 		this.snapshot = withSkip((expected, ...rest) => {
+			if (disableSnapshots && experiments.disableSnapshotsInHooks) {
+				fail(new AssertionError({
+					assertion: 'snapshot',
+					message: '`t.snapshot()` cannot be used here',
```
Comment (tymfear) [inline lib/assert.js:697]: If I understand it correctly, `t.snapshot()` and `t.try` are only allowed to be used in tests.
If that's so, I would suggest changing error to something like
```
`t.snapshot()` can be used in tests only
```

File: test-tap/hooks.js (hunk 3 above)
Comment (okyantoro) [reply test-tap/hooks.js]: Addressed. I still use same naming convention like `test-timeouts` for the directory, but we can change it if you prefer `hooks`.

Comment (okyantoro) [reply test-tap/hooks.js]: Addressed. I still use same naming convention like test-timeouts for the directory, but we can change it if you prefer hooks.

**Comment by tymfear** [comment]
@novemberborn when something is released under the feature flag, when it is a good time to update docs?
Changes in this PR should be reflected in [04-snapshot-testing.md](https://github.com/avajs/ava/blob/master/docs/04-snapshot-testing.md) and [03-assertions.md](https://github.com/avajs/ava/blob/master/docs/03-assertions.md#trytitle-implementation--macro--macro-args).

File: lib/assert.js (hunk 4 above)
Comment (novemberborn) [reply lib/assert.js]: Yes. Changing to "can only be used in tests"

**Comment by okyantoro** [comment]
Thank you @novemberborn

**Pull request merged**